WIP: Add projection split criteria (see issue #4)

morgsmss7 opened the pull request:
### What does this implement/fix? Explain your changes.
Fixes Issue #4. See also #2. 
#### Implemented the following split criteria:

- Axis projection: one predictor is chosen at random to calculate MSE
- Oblique projection: outputs are projected onto a sparse line before splitting with MSE

Note: These required bringing the random state into the RegressionCriterion class

#### Added basic tests for new split criteria in sklearn/tree/tests/tree/test_tree.py:
 - test_axis_proj: compares mse node impurity and child impurity calculations for a single predictor to the associated values for axis projection on a multi-output toy dataset

 - test_oblique_proj: compares mse node impurity and child impurity calculations for a single predictor or two predictors to the associated values for axis projection on a multi-output toy dataset

#### Demo Notebook
(in progress)

### Any other comments? (Especially seeking help/input for the following)
- Please note that the oblique projection test does not pass at the moment as I need to find a way to use try/except with assert statements.
- In creating tests, I also found it difficult to work with the random state, so I am suspicious of my addition of this parameter.
- I calculated MSE by hand for both tests, but did not end up using these calculations because they did not even match for the standard mse split criterion. I suspect it is due to rounding error, but I'm not sure.
- I could not figure out what rf.tree_.value.flat was, so I was unable to test this effectively. I commented out tests I tried.
- @v715 is working on finding out where the existing splitters fail. I am currently running an additional simulation that I expect to see the new split criteria excel at. If this works as expected, I will add this test to this pull request.

adam2392 commented on sklearn/tree/_criterion.pyx:
```diff
@@ -24,8 +24,11 @@ from libc.math cimport fabs
 
 import numpy as np
 cimport numpy as np
+import random #added by morgan
```
"#added by morgan" Are these necessary?

morgsmss7 replied:
No, they aren't. Thanks for reminding me. I was trying to keep track of lines I added incase issues arose later. I'll be sure to remove them.

adam2392 replied:
no problem! I know the sklearn folks will probably have you remove them anyways down the line. But a good tool to use for keeping track of your stuff you added is checking the PR "files changed", as well as using "git diff".

adam2392 commented on sklearn/tree/_criterion.pyx:
```diff
@@ -699,11 +701,17 @@ cdef class RegressionCriterion(Criterion):
         n_samples : SIZE_t
             The total number of samples to fit on
+
+        random_state : object #added by morgan
```
repeated in description.

adam2392 commented on sklearn/tree/_criterion.pyx:
```diff
@@ -734,7 +742,7 @@ cdef class RegressionCriterion(Criterion):
             raise MemoryError()
 
-    def __reduce__(self):
+    def __reduce__(self): #TODO do I need to add this for random_state
```
@bdpedigo what do you think?

morgsmss7 commented:
Hmm. Not sure why that is happening. I haven't tried `make`. I've been mainly using `make in` and that finishes without any issues. I always have to run this file before I can use the modified sklearn. Just be careful with the exports because I don't append things, just replace them because that was the only way I could get it to work.

```#!/bin/bash
export CC=/usr/bin/clang
export CXX=/usr/bin/clang++
export CPPFLAGS="-Xpreprocessor -fopenmp"
export CFLAGS="-I/usr/local/opt/libomp/include"
export CXXFLAGS="-I/usr/local/opt/libomp/include"
export LDFLAGS="-Wl,-rpath,/usr/local/opt/libomp/lib -L/usr/local/opt/libomp/lib -lomp"

make clean
pip install --verbose --editable .
make in

morgsmss7 commented:
@adam2392 I apologize for the delay. My laptop died causing me to lose a lot of my work from last week. I'm using a loaner right now, but I won't have it much longer as it is supposed to be my work laptop, and I need to wipe it to put a new OS on it. 

Anyway, I think I pretty much replicated what I had.
Here are the main changes I made:
1. Added `pred_weights` to the `SplitRecord` struct in order to share predictor weights across axis and oblique projection impurity calculations for a single node. (Previously, new weights were generated between calculating parent and child impurities, but that was not exactly what Vivek had in mind, so I implemented the shared weights)
2. Changed both axis and oblique impurity calculations to match the formula: 
`sum((y-mean)^2))`
I also changed `test_axis_proj` and `test_oblique_proj` to reflect this change.
3. Moved projection tests to their own file (`test_proj_criteria.py`)
These tests can be run using `python test_proj_criteria.py`. If "axis passed!" and "oblique passed!" are printed, both tests have passed without errors. Occasionally, I have been getting segfaults after these print statements, but I attribute this to something other than the tests themselves because it happens after all lines are executed.

morgsmss7 commented:
Because running Vivek's simulations resulted in many errors, I have decided to revert to unshared weights (a previous implementation that ran without errors). I have, however, updated the tests in `test_tree.py` and changed how MSE is calculated in Axis and Oblique split criteria to reflect by-hand calculations. I took out extraneous comments, and cleaned up the documentation. This commit should be good to go. I'll still be hunting for excess newlines, comments, etc., but other than that, this should be good.

morgsmss7 commented:
I'm not sure exactly the significance of these failed tests. They don't seem to pertain to anything I changed in this PR. Can a TA help me understand this a little better? Do I need to find a way to fix these before merging? @adam2392 @j1c @bdpedigo 

All pytests pass on my machine (understandably because they work on most of the MacOS tests). The failed tests seem to have something to do with missing data. I can't actually decipher what file causes the failure though.

morgsmss7 commented:
Actually, it looks like these tests were already failing in the latest commit to master.

<img width="1003" alt="image" src="https://user-images.githubusercontent.com/39866994/70855443-dd025000-1e98-11ea-8431-522408d744fe.png">

j1c commented:
> I'm not sure exactly the significance of these failed tests. They don't seem to pertain to anything I changed in this PR. Can a TA help me understand this a little better? Do I need to find a way to fix these before merging? @adam2392 @j1c @bdpedigo
> 
> All pytests pass on my machine (understandably because they work on most of the MacOS tests). The failed tests seem to have something to do with missing data. I can't actually decipher what file causes the failure though.

No you dont have to fix the failing builds. I just wanted to make sure new tests ran on _some_ builds..

morgsmss7 commented:
@adam2392 Is there something specific I should do so this is no longer a WIP?

adam2392 commented:
> @adam2392 Is there something specific I should do so this is no longer a WIP?

lol no sorry, this is just to tag it so I remind us to review it. 

I've skimmed it and it looks mostly fine rn.
